Signal/src/ViewControllers/HomeView/HomeViewController.m

```diff
@@ -1376,6 +1410,80 @@ - (void)updateEmptyBoxText
     _emptyBoxLabel.attributedText = fullLabelString;
 }
 
+- (void)showReRegistrationUI
+{
+    UIAlertController *actionSheetController =
+        [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
+
+    __weak HomeViewController *weakSelf = self;
+    [actionSheetController
+        addAction:[UIAlertAction
+                      actionWithTitle:NSLocalizedString(@"DEREGISTRATION_REREGISTER_WITH_SAME_PHONE_NUMBER",
+                                          @"Label for button that lets users re-register using the same phone number.")
+                                style:UIAlertActionStyleDestructive
+                              handler:^(UIAlertAction *action) {
+                                  [weakSelf reregisterWithSamePhoneNumber];
+                              }]];
+
+    [actionSheetController addAction:[OWSAlerts cancelAction]];
+
+    [self presentViewController:actionSheetController animated:YES completion:nil];
+}
+
+- (void)reregisterWithSamePhoneNumber
+{
+    DDLogInfo(@"%@ reregisterWithSamePhoneNumber.", self.logTag);
+
+    if (![[TSAccountManager sharedInstance] resetForReregistration]) {
+        OWSFail(@"%@ could not reset for re-registration.", self.logTag);
+        return;
+    }
+
+    [[Environment current].preferences unsetRecordedAPNSTokens];
+
+    [ModalActivityIndicatorViewController
```

Comment:
Arguably we could simplify this and just show the `RegistrationViewController` (which will be in re-registration mode) but this UI flow matches Signal Android.